### PR TITLE
Query generator for benchmarks

### PR DESCRIPTION
**Pull request by mrks**
If we want to run TPC-H with random values, we need to somehow generate queries. As a preparation, this PR changes all benchmarks to use an `AbstractQueryGenerator` instead of a vector of SQL strings.

Currently, we have two generators: One for TPC-H, and one for the file-based benchmark. For now, they simply return the queries the same way as before.

Also, this PR extracts the file-based benchmark from the benchmark runner, so that the interface becomes a bit cleaner.

### CODE REVIEW

File: src/benchmark/tpch_benchmark.cpp [hunk 1]
```diff
@@ -135,5 +131,5 @@ int main(int argc, char* argv[]) {
   context.emplace("scale_factor", scale_factor);
 
   // Run the benchmark
-  opossum::BenchmarkRunner(*config, queries, context).run();
+  opossum::BenchmarkRunner(*config, std::make_unique<opossum::TPCHQueryGenerator>(), context).run();
```
Comment (lanice) [inline src/benchmark/tpch_benchmark.cpp]: I think you forgot to pass the selected `query_ids` vector to the generator here.

Comment (mrks) [reply src/benchmark/tpch_benchmark.cpp]: fixed as well

File: src/benchmark/file_based_benchmark.cpp [hunk 2]
```diff
@@ -1,15 +1,68 @@
...
+  Assert(!tables.empty(), "No tables found in '" + table_path + "'");
+
+  for (const auto& table_path_str : tables) {
+    const auto table_name = filesystem::path{table_path_str}.stem().string();
+
+    std::shared_ptr<Table> table;
+    if (boost::algorithm::ends_with(table_path_str, ".tbl")) {
+      table = load_table(table_path_str, config.chunk_size);
+    } else {
+      table = CsvParser{}.parse(table_path_str);
```
[the remark this answers is not in the document]
Comment (lanice) [reply src/benchmark/file_based_benchmark.cpp]: Yep, sounds good :+1:

File: src/benchmark/CMakeLists.txt [hunk 3]
```diff
@@ -44,9 +44,16 @@ target_link_libraries(
 )
 
 # General purpose benchmark runner
-add_executable(hyriseBenchmark benchmark_runner_main.cpp)
+add_executable(
+    hyriseBenchmarkFiles
```
Comment (mrzzzrm) [inline src/benchmark/CMakeLists.txt]: That's not a good name. When I hear "hyriseBenchmarkFiles" I think of files describing a benchmark or something. Not of a benchmark executable that can load its data from disk. "hyriseBenchmark" was fine, imo. "hyriseBenchmarkRunner" would be okay with me as well.

File: src/benchmarklib/abstract_query_generator.hpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  virtual std::string build_query(const QueryID query_id) = 0;
+
+  // Returns the names of the individual queries (e.g., "TPC-H 1")
+  const std::vector<std::string>& query_names() const;
+
+  // Returns the number of queries supported by the benchmark
+  QueryID::base_type num_available_queries() const;
+
+  // Returns the number of queries selected for execution
+  QueryID::base_type num_selected_queries() const;
```
Comment (mrzzzrm) [inline src/benchmarklib/abstract_query_generator.hpp]: * Returning a ::base_type is not something we normally do. Why here? I'd just return size_t...
* In other places (Table::column_count(), ...) we use "_count" to indicate that the function returns a "count". Consistency would be nice here.

File: src/benchmark/CMakeLists.txt [hunk 5]
```diff
@@ -44,9 +44,16 @@ target_link_libraries(
...
+
+    file_based_benchmark.cpp
```
Comment (mrzzzrm) [inline src/benchmark/CMakeLists.txt]: "file_based" feels clunky to me. Why not just "generic" benchmark?

File: src/benchmarklib/tpch/tpch_query_generator.cpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+void TPCHQueryGenerator::_generate_names() {
+  _query_names.reserve(22);
+  for (auto i = 0; i < 22; ++i) {
+    _query_names.emplace_back(std::string("TPC-H ") + std::to_string(i + 1));
+  }
+}
+
+std::string TPCHQueryGenerator::build_query(const QueryID query_id) {
```
Comment (mrzzzrm) [inline src/benchmarklib/tpch/tpch_query_generator.cpp]: Just for my understanding: Once we support actual random query generation for TPC-H, this is the function that would return a randomly generated variant of the specified query?

File: src/benchmarklib/benchmark_utils.hpp [hunk 7]
```diff
@@ -37,16 +34,22 @@ using TableSegmentEncodingMapping =
  */
 std::ostream& get_out_stream(const bool verbose);
 
-struct QueryBenchmarkResult {
+struct QueryBenchmarkResult : public Noncopyable {
   QueryBenchmarkResult() { iteration_durations.reserve(1'000'000); }
+
+  QueryBenchmarkResult(QueryBenchmarkResult&& other) noexcept {
+    num_iterations.store(other.num_iterations);
+    duration = std::move(other.duration);
```
Comment (mrzzzrm) [inline src/benchmarklib/benchmark_utils.hpp]: move?

File: src/benchmarklib/tpch/tpch_query_generator.cpp (hunk 6 above)
Comment (mrks) [reply src/benchmarklib/tpch/tpch_query_generator.cpp]: Yes. Or, more specifically, an `EXECUTE` statement with the proper parameters.

File: src/benchmarklib/benchmark_utils.hpp (hunk 7 above)
Comment (mrks) [reply src/benchmarklib/benchmark_utils.hpp]: Well, it's a move constructor. If `move` doesn't do anything for this member, so be it. At least we're future-safe if someone decides to change the type of duration.

File: src/benchmarklib/abstract_query_generator.hpp (hunk 4 above)
Comment (mrks) [reply src/benchmarklib/abstract_query_generator.hpp]: ok

File: src/benchmark/CMakeLists.txt (hunk 5 above)
Comment (mrks) [reply src/benchmark/CMakeLists.txt]: "generic" doesn't convey any meaning to me.

Comment (mrzzzrm) [reply src/benchmark/CMakeLists.txt]: mmh, okay, if you think "file_based" is fine, then so be it

File: src/benchmarklib/tpch/tpch_query_generator.cpp (hunk 6 above)
Comment (mrzzzrm) [reply src/benchmarklib/tpch/tpch_query_generator.cpp]: And how would the "PREPARE" statements be generated?

File: src/benchmark/CMakeLists.txt (hunk 3 above)
Comment (mrks) [reply src/benchmark/CMakeLists.txt]: I'm somewhat trying to get away from the idea that the file-based runner is the "default" one (the way it was implemented before).

Comment (mrzzzrm) [reply src/benchmark/CMakeLists.txt]: I agree with you that there shouldn't be "primary" and "secondary" benchmarks. But "hyriseBenchmarkFiles" conveys a different meaning, to me at least, it sounds like "files that make up a benchmark". 
If "hyriseBenchmark" sounds too generic to you, why not "hyriseBenchmarkRunner"? From that name I, at least, could tell that it is an executable that **runs** a generic benchmark that I have specified. What's the generic way to specify a benchmark? Files with queries and files with tables.

Comment (lanice) [reply src/benchmark/CMakeLists.txt]: Hmm, but hyriseBenchmarkTPCH is a benchmarkRunner too, right?
How about "hyriseBenchmarkFileBased", similar to the actual file name?

Comment (mrzzzrm) [reply src/benchmark/CMakeLists.txt]: If "FileBased" is really a term we want to use, then, yeah, even that would be better than "hyriseBenchmarkFiles"

**Pull request merged**